packages/ampd-sdk/src/grpc/client.rs

```diff
@@ -55,103 +57,65 @@ pub trait Client {
-            multisig_prover: parse_addr(multisig_prover, "multisig prover")?,
-            service_registry: parse_addr(service_registry, "service registry")?,
-            rewards: parse_addr(rewards, "rewards contract")?,
-        })
+impl ManagedGrpcClient {
+    pub fn new(connection_handle: ConnectionHandle) -> Self {
+        Self { connection_handle }
     }
-}
-
-fn parse_addr(addr: &str, address_name: &'static str) -> Result<AccountId, Error> {
-    addr.parse::<AccountId>()
-        .change_context(AppError::InvalidAddress(address_name).into())
-        .attach_printable_lazy(|| addr.to_string())
-}
-
-pub enum KeyAlgorithm {
-    Ecdsa,
-    Ed25519,
-}
-
-pub struct Key {
-    pub id: nonempty::String,
-    pub algorithm: KeyAlgorithm,
-}
 
-impl From<Key> for KeyId {
-    fn from(key: Key) -> Self {
-        let algorithm = match key.algorithm {
-            KeyAlgorithm::Ecdsa => ampd_proto::Algorithm::Ecdsa,
-            KeyAlgorithm::Ed25519 => ampd_proto::Algorithm::Ed25519,
-        };
-
-        KeyId {
-            id: key.id.into(),
-            algorithm: algorithm as i32,
+    async fn get_service_clients(
+        &mut self,
+    ) -> Result<
+        (
+            BlockchainServiceClient<transport::Channel>,
+            CryptoServiceClient<transport::Channel>,
+        ),
+        Error,
+    > {
+        let connection_state = self.connection_handle.connection_receiver.borrow().clone();
+
+        match connection_state {
+            ConnectionState::Connected(channel) => {
+                let blockchain_service_client = BlockchainServiceClient::new(channel.clone());
+                let crypto_service_client = CryptoServiceClient::new(channel);
+                Ok((blockchain_service_client, crypto_service_client))
+            }
+            ConnectionState::Disconnected => {
+                bail!(Error::from(AppError::ConnectionUnavailable))
+            }
+            ConnectionState::Reconnecting => {
+                let mut receiver = self.connection_handle.connection_receiver.clone();
+                receiver.changed().await.into_report()?;
+                let connection_state = receiver.borrow().clone();
+
+                match connection_state {
+                    ConnectionState::Connected(channel) => {
+                        let blockchain_service_client =
+                            BlockchainServiceClient::new(channel.clone());
+                        let crypto_service_client = CryptoServiceClient::new(channel);
+                        Ok((blockchain_service_client, crypto_service_client))
+                    }
+                    _ => bail!(Error::from(AppError::ConnectionUnavailable)),
+                }
+            }
         }
     }
 }
 
 #[async_trait]
-impl Client for GrpcClient {
+impl Client for ManagedGrpcClient {
     type Stream = Pin<Box<dyn Stream<Item = Result<Event, Error>> + Send>>;
 
     async fn subscribe(
```

Comment:
this will still need the stream indirection, but that can be done in a follow up